lib/src/fixed_priority_scheduler.rs

```diff
@@ -0,0 +1,171 @@
+///
+/// # Arguments
+///
+/// * `processor` - An object that implements `ProcessorBase` trait, representing a CPU or a collection of CPU cores.
+/// * `task_list` - A mutable reference to a vector of NodeIndex, representing the tasks to be scheduled.
+/// * `dag` - A mutable reference to a Graph object, representing the dependencies among tasks.
+///
+/// # Returns
+///
+/// * A floating point number representing the normalized total time taken to finish all tasks.
+///
+/// # Description
+///
+/// The function first initializes a ready queue and a list of finished tasks, sets the time to 0,
+/// calculates the minimum_multiplier from the dag set and sets it to the processor.
+///
+/// Then, it enters a loop that continues until all tasks are finished.
+/// In each iteration, it finds the tasks whose predecessors have all finished and adds them to the ready queue.
+/// Afterward, it sorts the ready queue by the priority of tasks.
+///
+/// When there is an idle core, it assigns the first task in the ready queue to the idle core.
+///
+/// Then it processes tasks for one time unit and checks if there are tasks finished.
+/// If no tasks finish in this time unit, it continues processing tasks until there are tasks finished.
+///
+/// When a task finishes, it adds the task to the finished tasks list.
+///
+/// Finally, the function returns the total time taken to finish all tasks divided by the minimum_multiplier.
+///
+/// # Example
+///
+/// Refer to the examples in the tests code.
+///
+pub fn fixed_priority_scheduler(
+    processor: &mut impl ProcessorBase,
+    dag: &mut Graph<NodeData, i32>,
+) -> i32 {
+    let mut time = 0;
+    let mut ready_queue: VecDeque<NodeIndex> = VecDeque::new();
+    let mut finished_core_and_node_hashmap: HashMap<usize, NodeIndex> = HashMap::new();
+
+    let source_node = dag.add_dummy_source_node();
+    dag[source_node]
+        .params
+        .insert("execution_time".to_string(), DUMMY_EXECUTION_TIME);
+    let sink_node = dag.add_dummy_sink_node();
+    dag[sink_node]
+        .params
+        .insert("execution_time".to_string(), DUMMY_EXECUTION_TIME);
+
+    ready_queue.push_back(source_node);
+
+    loop {
+        //Sort by priority
+        ready_queue.make_contiguous().sort_by_key(|&node| {
+            let priority = dag[node].params.get("priority").unwrap_or(&999); //If there is no priority, it is set to the lowest priority.
+            *priority
+        });
+
+        //Assign the highest priority task first to the first idle core found.
+        while let Some(task) = ready_queue.pop_front() {
+            if let Some(core_index) = processor.get_idle_core_index() {
+                processor.allocate(core_index, dag[task].clone());
+                finished_core_and_node_hashmap.insert(core_index, task);
+            } else {
+                //If everything is working, it is useless to search any further, so the extracted task is returned to the ready queue.
```

Comment:
It is more efficient to check for idle cores before popping from ready_queue.

Reply:
while let Some(core_index) = processor.get_idle_core_index() {
    if let Some(task) = ready_queue.pop_front() {
        processor.allocate(core_index, dag[task].clone());
        finished_core_and_node_hashmap.insert(core_index, task);
    } else {
        break;
    }
}